ospfd/ospf_ext.c

```diff
@@ -0,0 +1,1777 @@
+	} else {
+		if (IS_DEBUG_OSPF_SR)
+			zlog_debug(
+				"EXT (ospf_ext_schedule_prefix_index) "
+				"Remove prefix for interface %s",
+				ifp->name);
+
+		if (CHECK_FLAG(exti->flags, EXT_LPFLG_LSA_ENGAGED)) {
+			ospf_ext_pref_lsa_schedule(exti, FLUSH_THIS_LSA);
+			UNSET_FLAG(exti->flags, EXT_LPFLG_LSA_ENGAGED);
+			UNSET_FLAG(exti->flags, EXT_LPFLG_LSA_ACTIVE);
+		}
+	}
+
+	return exti->instance;
+}
+
+/*
+ * Used by Segment Routing to activate/deactivate Extended Link/Prefix flooding
+ *
+ * @param enable To activate or not Segment Routing Extended LSA flooding
+ *
+ * @return none
+ */
+void ospf_ext_update_sr(bool enable)
+{
+	struct listnode *node;
+	struct ext_itf *exti;
+
+	if (IS_DEBUG_OSPF_SR)
+		zlog_debug(
+			"EXT (ospf_ext_update_sr): %s Extended LSAs for "
+			"Segment Routing",
+			enable ? "Enable" : "Disable");
+
+	if (enable) {
+		OspfEXT.enabled = true;
+		/* Refresh LSAs if already engaged or originate */
+		for (ALL_LIST_ELEMENTS_RO(OspfEXT.iflist, node, exti))
+			if (CHECK_FLAG(exti->flags, EXT_LPFLG_LSA_ENGAGED))
+				ospf_ext_lsa_schedule(exti, REFRESH_THIS_LSA);
+			else
+				ospf_ext_lsa_schedule(exti,
+						      REORIGINATE_THIS_LSA);
+	} else {
+		/* Start by Flushing engaged LSAs */
+		for (ALL_LIST_ELEMENTS_RO(OspfEXT.iflist, node, exti))
+			if (CHECK_FLAG(exti->flags, EXT_LPFLG_LSA_ENGAGED))
+				ospf_ext_lsa_schedule(exti, FLUSH_THIS_LSA);
+		/* And then disable Extended Link/Prefix */
+		OspfEXT.enabled = false;
+	}
+}
+/*------------------------------------------------------------------------*
+ * Followings are callback functions against generic Opaque-LSAs handling.
+ *------------------------------------------------------------------------*/
+
+/* Add new Interface in Extended Interface List */
+static int ospf_ext_link_new_if(struct interface *ifp)
+{
+	struct ext_itf *new;
+	int rc = -1;
+
+	if (lookup_ext_by_ifp(ifp) != NULL) {
+		zlog_warn(
+			"EXT (ospf_ext_link_new_if) interface %s"
+			" is already in use",
+			ifp ? ifp->name : "-");
+		rc = 0; /* Do nothing here. */
+		return rc;
+	}
+
+	new = XCALLOC(MTYPE_OSPF_EXT_PARAMS, sizeof(struct ext_itf));
+	if (new == NULL) {
+		zlog_warn("EXT: XCALLOC: %s", safe_strerror(errno));
+		return rc;
+	}
+
+	/* initialize new information and link back the interface */
+	new->ifp = ifp;
+	new->flags = EXT_LPFLG_LSA_INACTIVE;
+
+	listnode_add(OspfEXT.iflist, new);
+
+	rc = 0;
+	return rc;
+}
+
+/* Remove existing Interface from Extended Interface List */
+static int ospf_ext_link_del_if(struct interface *ifp)
+{
+	struct ext_itf *exti;
+	int rc = -1;
+
+	if ((exti = lookup_ext_by_ifp(ifp)) != NULL) {
+		struct list *iflist = OspfEXT.iflist;
+
+		/* Skip Extended Prefix interface */
+		if (exti->stype == PREF_SID)
+			return 0;
+
+		/* Dequeue listnode entry from the list. */
+		listnode_delete(iflist, exti);
+
+		/* Avoid misjudgement in the next lookup. */
+		if (listcount(iflist) == 0)
+			iflist->head = iflist->tail = NULL;
```

Comment:
If this is really necessary then our linked list implementation is terribly broken.

Reply:
agreed if we must be doing this, then either the implementation is broken or we've started playing with the internal data structures instead of using the accessor functions.